### DISCUSSION TITLE
I want to run Pyodide in an iOS offline environment.

### DISCUSSION THREAD

**Discussion by otmb** [pyodide/pyodide#2188] [Q&A] [answered]
I want to run Pyodide in an iOS offline environment.
HTTP Request worked fine online from WKWebView, but I get an error when fetching wasm with an offline local path.
`Unhandled Promise Rejection: TypeError: Unexpected response MIME type. Expected 'application/wasm'`

After some trial and error, I found that I could call wasm from XMLHttpRequest.
As a compromise, Overwrite fetch with XMLHttpRequest.

```
async function fetch(filename) {
    const promise = new Promise((resolve, reject) => {
        const xhr = new XMLHttpRequest();
        xhr.open('GET', filename);
        xhr.responseType = 'arraybuffer';
        xhr.addEventListener('load', (e) => resolve(xhr.response));
        xhr.send();
    });

    const response = new Response(await promise.then(), {headers: new Headers([['Content-Type', 'application/wasm']])});

    return response;
}
```

Is there any other good way?
Thanks.

Here is the code used for the evaluation.
https://gist.github.com/otmb/2eefc9249d347103469741542f135f5c

**Comment by rth** [comment, marked as answer]
> with an offline local path

Normally browsers will not allow loading .wasm files from local paths [see FAQ](https://pyodide.org/en/stable/usage/faq.html#why-can-t-i-load-files-from-the-local-file-system). I see you have 
```
        let config = WKWebViewConfiguration()
        config.preferences.setValue(true, forKey: "allowFileAccessFromFileURLs")
```
but I'm not sure if that is enough to bypass this restriction. 

In your case, I would start with launching a local HTTP server to serve Pyodide assets to make sure that works with WKWebView and Swift. Then once this works, try to get rid of the HTTP server to load files directly from local URLs, but I'm not sure that this would be technically possible.

**Reply by otmb** [reply]
Thank you for answering.

The specification that the local path cannot be read is helpful.
Even if I set `allowFileAccessFromFileURLs`, it seems that some files can be fetched and some cannot be read.
As it is said, it seems appropriate to have an HTTP server.
The XMLHttpRequest I've tried isn't mainstream as it can be blocked when the version changes.

**Comment by Klaus-Pete** [comment]
This one solved the problem for me: 
-> add a configuration to the info.plist in the iOS project:
https://github.com/ionic-team/capacitor/commit/7c0d58cc84472ceee26c96142122d9359efbe19c

**Reply by mbotsu** [reply]
capacitor uses a unique method to read files that does not require allowFileAccessFromFileURLs.
In order to advance the discussion here, it would be helpful if you could introduce the steps in Vanilla JS.
Thank.

**Reply by Klaus-Pete** [reply]
Sorry but my answer has nothing to do with “allowFileAccessFromFileURLs”. 
The error this thread started with has to do with the wrong MIME-type which is assigned to a file loaded by my application:
-- TypeError: Unexpected response MIME type. Expected 'application/wasm' --
The configuration described in https://github.com/ionic-team/capacitor/commit/7c0d58cc84472ceee26c96142122d9359efbe19c fixed that. A file with extension .wasm which is loaded by my app gets MIME-type application/wasm.
That way I do not need an extra local HTTP server.

**Reply by mbotsu** [reply]
OK. Looks like my explanation wasn't good enough.
As you say, wasm probably works with ionic-team/capacitor.
Instead, I think the point to be discussed is how ionic-team/capacitor loads wasm.

The proposed code includes the steps to embed WEBKIT in SwiftUI and run wasm on it.
https://gist.github.com/otmb/2eefc9249d347103469741542f135f5c

I think ionic-team/capacitor probably prepares a custom WEBKIT to achieve this.
The topic will be about the custom code.
